src/Sylius/Bundle/ThemeBundle/Asset/Installer/AssetsInstaller.php

```diff
@@ -0,0 +1,205 @@
+            // Do nothing, hard copy later.
+        }
+
+        $this->hardCopyAssets($originDir, $targetDir);
+
+        return self::HARD_COPY;
+    }
+
+    /**
+     * @param string $originDir
+     * @param string $targetDir
+     */
+    protected function relativeSymlinkAssets($originDir, $targetDir)
+    {
+        $this->doInstallDirAssets($originDir, $targetDir, true, true);
+    }
+
+    /**
+     * @param string $originDir
+     * @param string $targetDir
+     */
+    protected function symlinkAssets($originDir, $targetDir)
+    {
+        $this->doInstallDirAssets($originDir, $targetDir, true, false);
+    }
+
+    /**
+     * @param string $originDir
+     * @param string $targetDir
+     */
+    protected function hardCopyAssets($originDir, $targetDir)
+    {
+        $this->doInstallDirAssets($originDir, $targetDir, false, false);
+    }
+
+    /**
+     * @param string $originDir
+     * @param string $targetDir
+     * @param boolean $symlink
+     * @param boolean $relativeSymlink
+     */
+    protected function doInstallDirAssets($originDir, $targetDir, $symlink, $relativeSymlink)
+    {
+        $finder = new Finder();
+        $finder->sortByName()->ignoreDotFiles(false)->in($originDir);
+
+        $theme = $this->themeRepository->findByPath($originDir);
+
+        $this->filesystem->mkdir($targetDir);
+
+        /** @var SplFileInfo[] $finder */
+        foreach ($finder as $file) {
+            if ($file->isDir()) {
+                $this->filesystem->mkdir($targetDir . '/' . $file->getRelativePathname());
+            } else {
+                $targetFile = $targetDir . '/' . $file->getRelativePathname();
+                if (null !== $theme) {
+                    $targetFile = $this->pathResolver->resolve($targetFile, $theme);
+                }
+
+                $originFile = $file->getPathname();
+                if ($relativeSymlink) {
+                    $originFile = rtrim($this->filesystem->makePathRelative($originFile, realpath(dirname($targetFile))), '/');
+                }
+
+                if ($symlink) {
+                    $this->filesystem->symlink($originFile, $targetFile);
+
+                    if (!file_exists($targetFile)) {
+                        throw new IOException('Symbolic link is broken');
+                    }
+                } else {
+                    $this->filesystem->copy($originFile, $targetFile);
+                }
+            }
+        }
+    }
+
+    /**
+     * @param BundleInterface $bundle
+     *
+     * @return array
+     */
+    protected function findAssetsPathsForBundle(BundleInterface $bundle)
+    {
+        $sources[] = $bundle->getPath() . '/Resources/public';
+
+        foreach ($this->themeRepository->findAll() as $theme) {
+            $sources[] = $theme->getPath() . '/' . $bundle->getName() . '/public';
+        }
+
+        return array_filter($sources, function ($dir) {
+            return is_dir($dir);
+        });
```

Comment:
Why can't we filter dirs manually before we add them to array? Right now we iterate 2x through the array.

```php
$sourceDir = $theme->getPath() . '/' . $bundle->getName() . '/public';
if (is_dir($sourceDir)) {
   $sources[] =  $sourceDir;
}
```